EIPS/eip-3102.md

```diff
@@ -0,0 +1,189 @@
+| 4 | 3 * 12 = 36 | 1152 |
+| 8 | 7 * 8 = 56 | 1792 |
+| 16 | 15 * 6 = 90 | 2880 |
+
+Actual branch lengths will be slightly larger than this due to uneven distribution and overhead, but the pattern of `k=2` being by far the best remains.
+
+The ethereum tree was originally hexary because this would reduce the number of database reads (eg. 6 instead of 24 in the above example). It is now understood that this reasoning was flawed, because nodes can still "pretend" that a binary tree is a hexary (or even 256-ary) tree at the database layer (eg. see https://ethresear.ch/t/optimizing-sparse-merkle-trees/3751), and thereby get the best-of-both-worlds of having the low proof sizes of the tree being binary from a hash-structure perspective and at the same time a much more efficient representation in the database.
+
+Additionally, binary trees are expected to be widely used in Eth2, so this path improves forward-compatibility and reduces long-run total complexity for the protocol.
+
+### Path length instead of bit prefix
+
+In order to remove the complexity associated with byte manipulation, only the bit-length of the extension is used to merkelize a node with a prefix.
+
+Storing the length of the path from the root node instead of that from the parent node has the nice property that siblings need not be hashed when deleting a leaf.
+
+![Sibling deletion diagram](../assets/eip-3102/sibling.svg)
+
+_On the left, a trie with the prefix length, and on the right, a trie with the full path length. Each both have values `10000100` and `10000000`. Deleting `10000100` result in the same trie. In the case of the left trie, the sibling node has to be updated, while in the case of the right trie, it need not be._
+
+### Value hashing
+
+Except for the code, all values in the trie are less than 32 bytes. EIP-2926 introduces code chunks, with `CHUNK_SIZE = 32 bytes`. The hashing of the leaf's value could therefore be saved. The authors of the EIP are however considering a future increase of `CHUNK_SIZE`, making `hash(value)` the future-proof choice.
+
+## Backwards Compatibility
+
+A hard fork is required in order for blocks to have a trie root using a different structure.
+
+## Test Cases
+
+This section is still incomplete and will be fleshed out as clients start implementing the support for binary tries.
+
+An initial implementation has been made for Geth. Unit tests can be found [here](https://github.com/gballet/go-ethereum/blob/binary-trie-m5-full-path/trie/binary_test.go).
+
+## Implementation
+
+ * As of [commit 0db87e187dc0bfb96046a47e3d6768c93a2e3331](https://github.com/gballet/multiproof-rs/commit/6d22b1aef9548581826b3c04b3e00d6cc709388c), [multiproof-rs](https://github.com/gballet/multiproof-rs) implements this merkelization rule in the `hash_m5()` function, found in file `src/binary_tree.rs`.
+ * An implementation of this structure for go-ethereum is available [in this branch](https://github.com/gballet/go-ethereum/tree/binary-trie-m5).
+
```

Comment:
The implementation section is optional, and I general prefer that if it is included it has an inline reference implementation or something in `../assets/eip-3102/` rather than linking externally.  External links like this, especially to private repositories, tend to not stand the test of time very well.  Keep in mind that EIPs are meant to be pure technical specifications, and things like examples, tutorials, libraries, etc. should be handled elsewhere.